File: backend/nfe/src/model/nfe_cana_fornecimento_diario_model.py
from src import db
from src.model.nfe_cana_model import NfeCanaModel


class NfeCanaFornecimentoDiarioModel(db.Model):
    __tablename__ = 'nfe_cana_fornecimento_diario'

    id = db.Column(db.Integer, primary_key=True)
    dia = db.Column(db.String(2))
    quantidade = db.Column(db.Float)
    quantidade_total_mes = db.Column(db.Float)
    quantidade_total_anterior = db.Column(db.Float)
    quantidade_total_geral = db.Column(db.Float)
    id_nfe_cana = db.Column(db.Integer, db.ForeignKey('nfe_cana.id'))

    nfe_cana_model = db.relationship('NfeCanaModel', foreign_keys=[id_nfe_cana])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cana = data.get('idNfeCana')
        self.dia = data.get('dia')
        self.quantidade = data.get('quantidade')
        self.quantidade_total_mes = data.get('quantidadeTotalMes')
        self.quantidade_total_anterior = data.get('quantidadeTotalAnterior')
        self.quantidade_total_geral = data.get('quantidadeTotalGeral')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCana': self.id_nfe_cana,
            'dia': self.dia,
            'quantidade': self.quantidade,
            'quantidadeTotalMes': self.quantidade_total_mes,
            'quantidadeTotalAnterior': self.quantidade_total_anterior,
            'quantidadeTotalGeral': self.quantidade_total_geral,
            'nfeCanaModel': self.nfe_cana_model.serialize() if self.nfe_cana_model else None,
        }

File: frontend/nfe/lib/app/data/model/nfe_cana_fornecimento_diario_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/data/model/model_imports.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeCanaFornecimentoDiarioModel {
	int? id;
	int? idNfeCana;
	String? dia;
	double? quantidade;
	double? quantidadeTotalMes;
	double? quantidadeTotalAnterior;
	double? quantidadeTotalGeral;
	NfeCanaModel? nfeCanaModel;

	NfeCanaFornecimentoDiarioModel({
		this.id,
		this.idNfeCana,
		this.dia,
		this.quantidade,
		this.quantidadeTotalMes,
		this.quantidadeTotalAnterior,
		this.quantidadeTotalGeral,
		this.nfeCanaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'dia',
		'quantidade',
		'quantidade_total_mes',
		'quantidade_total_anterior',
		'quantidade_total_geral',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Dia',
		'Quantidade',
		'Quantidade Total Mes',
		'Quantidade Total Anterior',
		'Quantidade Total Geral',
	];

	NfeCanaFornecimentoDiarioModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCana = jsonData['idNfeCana'];
		dia = NfeCanaFornecimentoDiarioDomain.getDia(jsonData['dia']);
		quantidade = jsonData['quantidade']?.toDouble();
		quantidadeTotalMes = jsonData['quantidadeTotalMes']?.toDouble();
		quantidadeTotalAnterior = jsonData['quantidadeTotalAnterior']?.toDouble();
		quantidadeTotalGeral = jsonData['quantidadeTotalGeral']?.toDouble();
		nfeCanaModel = jsonData['nfeCanaModel'] == null ? NfeCanaModel() : NfeCanaModel.fromJson(jsonData['nfeCanaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCana'] = idNfeCana != 0 ? idNfeCana : null;
		jsonData['dia'] = NfeCanaFornecimentoDiarioDomain.setDia(dia);
		jsonData['quantidade'] = quantidade;
		jsonData['quantidadeTotalMes'] = quantidadeTotalMes;
		jsonData['quantidadeTotalAnterior'] = quantidadeTotalAnterior;
		jsonData['quantidadeTotalGeral'] = quantidadeTotalGeral;
		jsonData['nfeCanaModel'] = nfeCanaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCana = plutoRow.cells['idNfeCana']?.value;
		dia = plutoRow.cells['dia']?.value != '' ? plutoRow.cells['dia']?.value : 'AAA';
		quantidade = plutoRow.cells['quantidade']?.value?.toDouble();
		quantidadeTotalMes = plutoRow.cells['quantidadeTotalMes']?.value?.toDouble();
		quantidadeTotalAnterior = plutoRow.cells['quantidadeTotalAnterior']?.value?.toDouble();
		quantidadeTotalGeral = plutoRow.cells['quantidadeTotalGeral']?.value?.toDouble();
		nfeCanaModel = NfeCanaModel();
		nfeCanaModel?.safra = plutoRow.cells['nfeCanaModel']?.value;
	}	

	NfeCanaFornecimentoDiarioModel clone() {
		return NfeCanaFornecimentoDiarioModel(
			id: id,
			idNfeCana: idNfeCana,
			dia: dia,
			quantidade: quantidade,
			quantidadeTotalMes: quantidadeTotalMes,
			quantidadeTotalAnterior: quantidadeTotalAnterior,
			quantidadeTotalGeral: quantidadeTotalGeral,
		);			
	}

	
}